azurerm/resource_arm_policy_set_definition.go

```diff
@@ -0,0 +1,283 @@
+			},
+		},
+	}
+}
+
+func policyDefinitionsDiffSuppressFunc(k, old, new string, d *schema.ResourceData) bool {
+	var oldPolicyDefinitions []policy.DefinitionReference
+	errOld := json.Unmarshal([]byte(old), &oldPolicyDefinitions)
+	if errOld != nil {
+		return false
+	}
+
+	var newPolicyDefinitions []policy.DefinitionReference
+	errNew := json.Unmarshal([]byte(old), &newPolicyDefinitions)
+	if errNew != nil {
+		return false
+	}
+
+	return reflect.DeepEqual(oldPolicyDefinitions, newPolicyDefinitions)
+}
+
+func resourceArmPolicySetDefinitionCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).policySetDefinitionsClient
+	ctx := meta.(*ArmClient).StopContext
+
+	name := d.Get("name").(string)
+	policyType := d.Get("policy_type").(string)
+	displayName := d.Get("display_name").(string)
+	description := d.Get("description").(string)
+
+	properties := policy.SetDefinitionProperties{
+		PolicyType:  policy.Type(policyType),
+		DisplayName: utils.String(displayName),
+		Description: utils.String(description),
+	}
+
+	if metaDataString := d.Get("metadata").(string); metaDataString != "" {
+		metaData, err := structure.ExpandJsonFromString(metaDataString)
+		if err != nil {
+			return fmt.Errorf("unable to expand metadata json: %s", err)
+		}
+		properties.Metadata = &metaData
+	}
+
+	if parametersString := d.Get("parameters").(string); parametersString != "" {
+		parameters, err := structure.ExpandJsonFromString(parametersString)
+		if err != nil {
+			return fmt.Errorf("unable to expand parameters json: %s", err)
+		}
+		properties.Parameters = &parameters
+	}
+
+	if policyDefinitionsString := d.Get("policy_definitions").(string); policyDefinitionsString != "" {
+		var policyDefinitions []policy.DefinitionReference
+		err := json.Unmarshal([]byte(policyDefinitionsString), &policyDefinitions)
+		if err != nil {
+			return fmt.Errorf("unable to expand parameters json: %s", err)
+		}
+		properties.PolicyDefinitions = &policyDefinitions
+	}
+
+	definition := policy.SetDefinition{
+		Name:                    utils.String(name),
+		SetDefinitionProperties: &properties,
+	}
+
+	if _, err := client.CreateOrUpdate(ctx, name, definition); err != nil {
+		return err
+	}
+
+	// Policy Definitions are eventually consistent; wait for them to stabilize
+	log.Printf("[DEBUG] Waiting for Policy Set Definition %q to become available", name)
+	stateConf := &resource.StateChangeConf{
+		Pending:                   []string{"404"},
+		Target:                    []string{"200"},
+		Refresh:                   policySetDefinitionRefreshFunc(ctx, client, name),
+		Timeout:                   5 * time.Minute,
+		MinTimeout:                10 * time.Second,
+		ContinuousTargetOccurence: 10,
+	}
+	if _, err := stateConf.WaitForState(); err != nil {
+		return fmt.Errorf("Error waiting for Policy Set Definition %q to become available: %s", name, err)
+	}
+
+	resp, err := client.Get(ctx, name)
+	if err != nil {
+		return err
```

Comment:
(as above) could we wrap this error to display a more useful error:

```suggestion
		return fmt.Errorf("Error retrieving Policy Set Definition %q: %s", name, err)
```